### ISSUE TITLE
Expected SSH_FXP_ATTRS or SSH_FXP_STATUS 

### ISSUE THREAD

**Issue by narem81** [phpseclib/phpseclib#999]
Hi everybody,
I'm using this great library for creating a Xen virtual machine on a remote machine. This involves the execution of xen-create-image command. The execution of this command lasts around 10 minutes if launched on a local shell (on the machine where Xen is installed). 
If launched remotely, through an SFTP class object, the execution of the xen-create-image lasts the same, however SFTP::exec method returns much earlier of it (less than 1 minute).

When xen-create-image ends its execution, a notification displays the message "Expected SSH_FXP_ATTRS or SSH_FXP_STATUS" and the php code stops its execution showing a read_user() call error.

What this problem is related with?
Any hints on how to solve it?

Thanks in advance.

**Comment by terrafrost** [comment]
What version of phpseclib are you using? Can you do `define('NET_SSH2_LOGGING', 2)` at the top and then do `$sftp->getLog()` after you get the error?

Thanks!

**Comment by narem81** [comment]
@terrafrost 
I'm using phpseclib version 2.0.2. 
I updated to this version since I had the same problem with version 2.0.0. 
I used logging and did var_dump($sftp->getLog()). 
This is what I got:
`string(40280) "<-
00000000  53:53:48:2d:32:2e:30:2d:4f:70:65:6e:53:53:48:5f  SSH-2.0-OpenSSH_
00000010  36:2e:36:2e:31:70:31:20:55:62:75:6e:74:75:2d:32  6.6.1p1 Ubuntu-2
00000020  75:62:75:6e:74:75:32:0d:0a                       ubuntu2..

->
00000000  53:53:48:2d:32:2e:30:2d:70:68:70:73:65:63:6c:69  SSH-2.0-phpsecli
00000010  62:5f:32:2e:30:20:28:6f:70:65:6e:73:73:6c:2c:20  b_2.0 (openssl, 
00000020  62:63:6d:61:74:68:29:0d:0a                       bcmath)..

<- NET_SSH2_MSG_KEXINIT (since last: 0.0198, network"...
`
Thank you very much for your help :)

**Comment by terrafrost** [comment]
Can you post the full string? It looks like you've truncated what you've posted.

If it's too big try posting it on pastebin.com and then posting the link here.

Thanks!

**Comment by narem81** [comment]
Ok, this is the link : 
http://pastebin.com/4PCcsABG

I'm also uploading it as a file:
[sftp.txt](https://github.com/phpseclib/phpseclib/files/318778/sftp.txt)

Thanks again!

**Comment by narem81** [comment]
If this could be helpful, the xen-create-image command is a wrapper of three commands, executed sequentially. What I noticed is that SFTP::exec stops its execution after the first built-in command aproximately.

**Comment by terrafrost** [comment]
Can you post your code?

It looks like you're first stat'ing /etc/xen/vms and then /etc/xen/auto. You then run "mv /etc/xen-tools/xen-tools.conf /etc/xen-tools/xen-tools-base.conf".

You then upload a new /etc/xen-tools/xen-tools.conf and then you run "xen-create-image --hostname=netserver1 --ip=192.168.25.173 --genpass=0 --verbose" and then you get a bunch of data back.

The "Expected SSH_FXP_ATTRS or SSH_FXP_STATUS" is "thrown" by _stat (which is called by all the functions that do any sort of file attribute stuff) and chmod(). exec() doesn't "throw" that error. It's not defined in SFTP.php (just SSH2.php) and the stirng "SSH_FXP" does not occur in SSH2.php (nor does the string "n exec" in SFTP.php).

If I didn't know better I'd guess that I wasn't seen the whole lot file. Maybe if you do `define('NET_SSH2_LOGGING', 3)`, which'll enable real-time logging and give me the output of that I'd have better success in figuring out what's going on.

Thanks!

**Comment by narem81** [comment]
Ok, this is the new log produced by SFTP  class, version 2.0.0, comprehensive of Exception messages.

[sftplog.txt](https://github.com/phpseclib/phpseclib/files/323838/sftplog.txt)

This is the piece of code running the command:
`if ($sftp->file_exists("/etc/xen-tools/xen-tools.conf"))
            $sftp->exec("mv /etc/xen-tools/xen-tools.conf /etc/xen-tools/xen-tools-base.conf");
        $sftp->put("/etc/xen-tools/xen-tools.conf", $xt_conf);
        Logger::infolog("Xen: Creazione file di configurazione di xen tools completata");
        Logger::infolog("Xen-tools: Creazione macchina virtuale ".$vmname);
        $command = "xen-create-image --hostname=".$vmname." --ip=".$preconf["address"];
        $command .= " --genpass=0 --password=".$vmcommon["rootpwd"]. " --noverbose";
        $output = "";
        $count = 0;
        //$sftp->enablePTY();
        //$sftp->write($command."\n");
        $output = $sftp->exec($command);
        /*do
        {
            $output .= $sftp->read();
            $count++;
        }while(strpos($output,"Summary")===false && $count<100);*/

        while(!$sftp->file_exists("/etc/xen/vms/" . $vmname . ".cfg") && $count < 40) {
            if($count%2 === 0) Logger::infolog("Attendo la creazione del file...Tempo trascorso: ".$count." minuti");
            $count += 1;
            sleep(60);
        }

        echo `"\n\n$output\n\n";`

Thank you again for your help!

**Comment by terrafrost** [comment]
1. > I need to deploy the code on a production environment, therefore I'm using 2.0.0 version now as it's the last stable version (the 2.0.2 threw 70ish errors when I committed the project). 

  What were the errors? I don't get any when I do this:

  ```php
  define('NET_SSH2_LOGGING', 3);

  use \phpseclib\Net\SFTP;

  $sftp = new SFTP('domain.tld');
  $sftp->login('user', 'pass');

  $sftp->nlist();
  $sftp->exec('ls -latr');
  ```

2. From your sftplog.txt...

  ```
  <- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0002, network: 0s)
  00000000  00:00:00:00:00:00:00:7b:44:6f:6e:65:0a:0a:4e:6f  .......{Done..No
  00000010  20:72:6f:6c:65:20:73:63:72:69:70:74:73:20:77:65   role scripts we
  00000020  72:65:20:73:70:65:63:69:66:69:65:64:2e:20:20:53  re specified.  S
  00000030  6b:69:70:70:69:6e:67:0a:53:65:74:74:69:6e:67:20  kipping.Setting 
  00000040  75:70:20:72:6f:6f:74:20:70:61:73:73:77:6f:72:64  up root password
  00000050  0a:47:65:6e:65:72:61:74:69:6e:67:20:61:20:70:61  .Generating a pa
  00000060  73:73:77:6f:72:64:20:66:6f:72:20:74:68:65:20:6e  ssword for the n
  00000070  65:77:20:67:75:65:73:74:2e:0a:41:6c:6c:20:64:6f  ew guest..All do
  PHP Notice:  Expected SSH_FXP_ATTRS or SSH_FXP_STATUS in /home/cfortunato/PhpstormProjects/nosp_automaker/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 1282
  00000080  6e:65:0a                                         ne.
  ```

  The fact that the PHP Notice is appearing in the middle of the log output is very strange. `_append_log`, as defined in SSH2.php, outputs the log files and it doesn't "throw" errors in the middle of it's execution. Not like the one you're seeing.

  Normally, I might think there were some `\r`'s being thrown around but if that were the case text would be replaced and that's not what's happening in your code - you have the PHP Notice appearing smack dab between the `00000070` line and the `00000080` line.

3. In SSH2.php (`_get_channel_packet`) there's this:

  ```php
                    if ($client_channel == $channel) {
                        return $data;
                    }
                    if (!isset($this->channel_buffers[$channel])) {
                        $this->channel_buffers[$channel] = array();
                    }
                    $this->channel_buffers[$channel][] = $data;
                    break;
  ```
  That's in a while loop. Until it reaches data on the desired channel it just keeps looping. But in your case it's like it's receiving data for the exec channel, thinking it's for the SFTP channel, and then returning it to SFTP's `_stat` method, which is causing the error that you're seeing. Only problem: I don't see how this is possible. But, regardless, to give me more info with which a diagnosis can be made...  do this:

  ```
  #
  #-----[ OPEN ]------------------------------------------
  #
  Net/SSH2.php
  #
  #-----[ FIND ]------------------------------------------
  #
  if ($client_channel == $channel) {
  #
  #-----[ BEFORE, ADD ]-----------------------------------
  #
  echo "$client_channel == $channel\n";
  ```
  So apply that change, rerun your script (with logging set to 3), and get me the new output.

**Comment by narem81** [comment]
Good morning,
in order to let you focus on the code, I made a simplified script (keeping only the essential code) and I ran it with your indications. 
This is the code:

```
<?php
include('../vendor/autoload.php');

use phpseclib\Net\SFTP;

define('NET_SSH2_LOGGING', 3);

$sftp = new SFTP('192.168.25.171');
exec("ssh-keygen -R 192.168.25.171");
if (!$sftp->login('root', 'nvcm-lab')) {
exit('Login Failed');
}

if (!$sftp->file_exists("/etc/xen/vms")) $sftp->exec("mkdir /etc/xen/vms");
if (!$sftp->file_exists("/etc/xen/auto")) $sftp->exec("mkdir /etc/xen/auto");

$xt_conf = "";
$xt_conf .= "lvm = vg-systems";
$xt_conf .= "\ninstall-method = debootstrap";
$xt_conf .= "\nsize = 10G";
$xt_conf .= "\nmemory = 1G";
$xt_conf .= "\nnoswap = 1";
$xt_conf .= "\nfs = ext4";
$xt_conf .= "\ndist = `xt-guess-suite-and-mirror --suite`";
$xt_conf .= "\nimage = sparse";
$xt_conf .= "\nbridge = bridge1";
$xt_conf .= "\ngateway = 192.168.25.254";
$xt_conf .= "\nnetmask = 255.255.255.0";
$xt_conf .= "\nnameserver = 192.168.25.253";
$xt_conf .= "\nkernel = /boot/vmlinuz-`uname -r`";
$xt_conf .= "\ninitrd = /boot/initrd.img-`uname -r`";
$xt_conf .= "\npygrub = 1";
$xt_conf .= "\narch = amd64";
$xt_conf .= "\nmirror = `xt-guess-suite-and-mirror --mirror`";
$xt_conf .= "\next4_options     = noatime,nodiratime,errors=remount-ro";
$xt_conf .= "\next3_options     = noatime,nodiratime,errors=remount-ro";
$xt_conf .= "\next2_options     = noatime,nodiratime,errors=remount-ro";
$xt_conf .= "\nxfs_options      = defaults";
$xt_conf .= "\nreiserfs_options = defaults";
$xt_conf .= "\nbtrfs_options    = defaults";
$xt_conf .= "\noutput    = /etc/xen/vms";

if ($sftp->file_exists("/etc/xen-tools/xen-tools.conf"))
    $sftp->exec("mv /etc/xen-tools/xen-tools.conf /etc/xen-tools/xen-tools-base.conf");
$sftp->put("/etc/xen-tools/xen-tools.conf", $xt_conf);

$command = "xen-create-image --hostname=trialMac --ip=192.168.25.177";
$command .= " --genpass=0 --password=trialpwd --noverbose";
$output = "";

$count = 0;

$output = $sftp->exec($command);

while(!$sftp->file_exists("/etc/xen/vms/trialMac.cfg")) {
    if($count%2 === 0) echo("Attendo la creazione del file...Tempo trascorso: ".$count." minuti\n");
    $count += 1;
    sleep(60);
}

echo "\n\n$output\n\n";
$sftplog = $sftp->getLog();
file_put_contents("sftplog.txt",$sftplog);

$sftp->disconnect();


?>
```
And here is the generated log:
[sftplog_new.txt](https://github.com/phpseclib/phpseclib/files/327179/sftplog_new.txt)

Hoping this could make things clear, thanks again.

**Comment by narem81** [comment]
As you may notice from the code above, I put a while loop after the exec method call, just because that method returns too early, before the completion of the xen-create-image command, which seems odd to me.

**Comment by terrafrost** [comment]
Actually, let's try this:

```
#
#-----[ OPEN ]------------------------------------------
#
Net/SFTP.php
#
#-----[ FIND ]------------------------------------------
#
user_error('Expected SSH_FXP_ATTRS or SSH_FXP_STATUS');
#
#-----[ BEFORE, ADD ]-----------------------------------
#
echo "packet type = {$this->packet_type}\n";
echo bin2hex($response) . "\n"
```

Thanks!

**Comment by narem81** [comment]
Hi, I added the echo statement you wrote and here is the generated log:

[sftplog_20160623.txt](https://github.com/phpseclib/phpseclib/files/330094/sftplog_20160623.txt)

Looking at the log, as far as I can see, no echo message has been written.

The code I ran is always the same as that I posted yesterday.

**Comment by narem81** [comment]
Sorry, I put the statements before only one user_error, just inside the get method... my fault. Have I to add them before all user_error calls or just inside one specific method ?

**Comment by narem81** [comment]
Ok, I've done the test again.
This is the log: 

[sftplog_20160623_new.txt](https://github.com/phpseclib/phpseclib/files/330235/sftplog_20160623_new.txt)

So don't consider the previous log , named sftplog_20160623.txt, but instead consider this new one.
Thanks

**Comment by terrafrost** [comment]
Something that's pretty messed up... so per the code change I gave you you'd think that `packet_type` would appear in the output **before** the "PHP Notice". But, in fact, it's not. It appears after:

```
<- NET_SSH2_MSG_CHANNEL_REQUEST (since last: 0.0006, network: 0s)
00000000  00:00:00:00:00:00:00:0b:65:78:69:74:2d:73:74:61  ........exit-sta
00000010  74:75:73:00:00:00:00:00                          tus.....



PHP Notice:  Expected SSH_FXP_ATTRS or SSH_FXP_STATUS in /home/cfortunato/PhpstormProjects/nosp_automaker/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 1285
<- NET_SSH2_MSG_CHANNEL_EOF (since last: 0.0001, network: 0s)
00000000  00:00:00:00                                      ....



<- NET_SSH2_MSG_CHANNEL_CLOSE (since last: 0.0001, network: 0s)
00000000  00:00:00:00                                      ....



-> NET_SSH2_MSG_CHANNEL_CLOSE (since last: 0.0001, network: 0s)
00000000  00:00:00:01                                      ....


packet type = 


-> NET_SSH2_MSG_CHANNEL_DATA (since last: 60.0003, network: 0s)
00000000  00:00:00:00:00:00:00:26:00:00:00:22:11:00:00:00  .......&..."....
00000010  01:00:00:00:19:2f:65:74:63:2f:78:65:6e:2f:76:6d  ...../etc/xen/vm
00000020  73:2f:74:72:69:61:6c:4d:61:63:2e:63:66:67        s/trialMac.cfg
```

So that's pretty messed up and defies logic.

Anyway, try this:

```
#
#-----[ OPEN ]------------------------------------------
#
Net/SFTP.php
#
#-----[ FIND ]------------------------------------------
#
            if (is_bool($temp)) {
                $this->packet_type = false;
                $this->packet_buffer = '';
                return false;
            }
#
#-----[ REPLACE WITH ]----------------------------------
#
            if (is_bool($temp)) {
echo "_get_channel_packet returned a boolean\n";
                $this->packet_type = false;
                $this->packet_buffer = '';
                return false;
            }
#
#-----[ OPEN ]------------------------------------------
#
Net/SSH2.php
#
#-----[ FIND ]------------------------------------------
# in _get_channel_packet
#
                if ($this->curTimeout < 0) {
                    $this->is_timeout = true;
                    return true;
                }
#
#-----[ REPLACE WITH ]----------------------------------
#
                if ($this->curTimeout < 0) {
echo "zzz x1\n";
                    $this->is_timeout = true;
                    return true;
                }
#
#-----[ FIND ]------------------------------------------
# in _get_channel_packet
#
                if (!@stream_select($read, $write, $except, $sec, $usec) && !count($read)) {
                    $this->is_timeout = true;
                    return true;
                }
#
#-----[ REPLACE WITH ]----------------------------------
#
                if (!@stream_select($read, $write, $except, $sec, $usec) && !count($read)) {
echo "zzz x2\n";
                    $this->is_timeout = true;
                    return true;
                }

**Comment by narem81** [comment]
Hello,
@terrafrost : I did what you asked and I ran the code (always the same code as that I posted 5 days ago). 

This is the log produced:
[sftplog_20160627.txt](https://github.com/phpseclib/phpseclib/files/334499/sftplog_20160627.txt)

I really appreciate your support.

**Comment by terrafrost** [comment]
From this latest log file:

```
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0009, network: 0.0007s)
00000000  00:00:01:00:00:00:00:21:00:00:00:1d:65:00:00:00  .......!....e...
00000010  01:00:00:00:02:00:00:00:0c:4e:6f:20:73:75:63:68  .........No such
00000020  20:66:69:6c:65:00:00:00:00                        file....


256 == 256
Attendo la creazione del file...Tempo trascorso: 6 minuti
PHP Notice:  Expected SSH_FXP_ATTRS or SSH_FXP_STATUS in /home/cfortunato/PhpstormProjects/nosp_automaker/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 1283

-> NET_SSH2_MSG_CHANNEL_DATA (since last: 60.0004, network: 0s)
00000000  00:00:00:00:00:00:00:26:00:00:00:22:11:00:00:00  .......&..."....
00000010  01:00:00:00:19:2f:65:74:63:2f:78:65:6e:2f:76:6d  ...../etc/xen/vm
00000020  73:2f:74:72:69:61:6c:4d:61:63:2e:63:66:67        s/trialMac.cfg
```

So you're getting the PHP Notice between an incoming NET_SSH2_MSG_CHANNEL_DATA packet and an outgoing one.

Yet in your [last logs](https://github.com/phpseclib/phpseclib/issues/999#issuecomment-228561703) you were getting it between an incoming NET_SSH2_MSG_CHANNEL_REQUEST and an incoming NET_SSH2_MSG_CHANNEL_EOF.

And then in the [logs before that](https://github.com/phpseclib/phpseclib/issues/999#issuecomment-227422856) you were getting that error in the *middle* of an incoming NET_SSH2_MSG_CHANNEL_DATA packet.

It doesn't make any sense at all and it is complicating my ability to actually diagnose this issue.

Anyway, let's try this:

```
#
#-----[ OPEN ]------------------------------------------
#
Net/SFTP.php
#
#-----[ FIND ]------------------------------------------
#
user_error('Expected SSH_FXP_ATTRS or SSH_FXP_STATUS');
#
#-----[ REPLACE WITH ]----------------------------------
#
echo "Expected SSH_FXP_ATTRS or SSH_FXP_STATUS\n";
#
#-----[ FIND ]------------------------------------------
#
user_error('Expected SSH_FXP_ATTRS or SSH_FXP_STATUS');
#
#-----[ REPLACE WITH ]----------------------------------
#
echo "Expected SSH_FXP_ATTRS or SSH_FXP_STATUS\n";
```

**Comment by narem81** [comment]
Good morning,
I don't know the reason of this odd behaviour. I guess that, as the SSH2::exec() method returns prematurely (before the xen-create-image command ends its execution) to the caller function code, the execution remains in the while loop until the cfg file is created. From my understanding, when the xen-create-image command finally creates that file and stops its execution, then that message shows up.
Anyway I didn't make any change to the code except what you asked me to do. From these latest changes you asked I ran the code and this is the produced log:
[sftplog_20160628.txt](https://github.com/phpseclib/phpseclib/files/336541/sftplog_20160628.txt)

Just to remember, this is the code I run:

```
<?php
include('../vendor/autoload.php');

use phpseclib\Net\SFTP;

define('NET_SSH2_LOGGING', 3);

$sftp = new SFTP('192.168.25.171');
exec("ssh-keygen -R 192.168.25.171");
if (!$sftp->login('root', 'nvcm-lab')) {
exit('Login Failed');
}

if (!$sftp->file_exists("/etc/xen/vms")) $sftp->exec("mkdir /etc/xen/vms");
if (!$sftp->file_exists("/etc/xen/auto")) $sftp->exec("mkdir /etc/xen/auto");

$xt_conf = "";
$xt_conf .= "lvm = vg-systems";
$xt_conf .= "\ninstall-method = debootstrap";
$xt_conf .= "\nsize = 10G";
$xt_conf .= "\nmemory = 1G";
$xt_conf .= "\nnoswap = 1";
$xt_conf .= "\nfs = ext4";
$xt_conf .= "\ndist = `xt-guess-suite-and-mirror --suite`";
$xt_conf .= "\nimage = sparse";
$xt_conf .= "\nbridge = bridge1";
$xt_conf .= "\ngateway = 192.168.25.254";
$xt_conf .= "\nnetmask = 255.255.255.0";
$xt_conf .= "\nnameserver = 192.168.25.253";
$xt_conf .= "\nkernel = /boot/vmlinuz-`uname -r`";
$xt_conf .= "\ninitrd = /boot/initrd.img-`uname -r`";
$xt_conf .= "\npygrub = 1";
$xt_conf .= "\narch = amd64";
$xt_conf .= "\nmirror = `xt-guess-suite-and-mirror --mirror`";
$xt_conf .= "\next4_options     = noatime,nodiratime,errors=remount-ro";
$xt_conf .= "\next3_options     = noatime,nodiratime,errors=remount-ro";
$xt_conf .= "\next2_options     = noatime,nodiratime,errors=remount-ro";
$xt_conf .= "\nxfs_options      = defaults";
$xt_conf .= "\nreiserfs_options = defaults";
$xt_conf .= "\nbtrfs_options    = defaults";
$xt_conf .= "\noutput    = /etc/xen/vms";

if ($sftp->file_exists("/etc/xen-tools/xen-tools.conf"))
    $sftp->exec("mv /etc/xen-tools/xen-tools.conf /etc/xen-tools/xen-tools-base.conf");
$sftp->put("/etc/xen-tools/xen-tools.conf", $xt_conf);

$command = "xen-create-image --hostname=trialMac --ip=192.168.25.177";
$command .= " --genpass=0 --password=trialpwd --noverbose";
$output = "";

$count = 0;

$output = $sftp->exec($command);

while(!$sftp->file_exists("/etc/xen/vms/trialMac.cfg")) {
    if($count%2 === 0) echo("Attendo la creazione del file...Tempo trascorso: ".$count." minuti\n");
    $count += 1;
    sleep(60);
}

//echo "\n\n$output\n\n";
//$sftplog = $sftp->getLog();
//file_put_contents("sftplog.txt",$sftplog);

$sftp->disconnect();


?>

```

**Comment by terrafrost** [comment]
That really helped - thanks! #1006 should fix the issue for you, I believe!

**Comment by narem81** [comment]
Thanks for the fix!
It removed the PHP Notice: "Expected SSH_FXP_ATTRS or SSH_FXP_STATUS" and that's good.

However it didn't prevent the SSH2::exec method to return prematurely. 

One quick note to hopefully clarify a little further based on the excerpt below.
```
1. $command = "xen-create-image --hostname=trialMac --ip=192.168.25.177";
2. $command .= " --genpass=0 --password=trialpwd --noverbose";
3. $output = $sftp->exec($command);
4. while(!$sftp->file_exists("/etc/xen/vms/trialMac.cfg")) {
        if($count%2 === 0) echo("Attendo la creazione del file...Tempo trascorso: ".$count." minuti\n");
        $count += 1;
        sleep(60);
   }

```
- The **expected** behaviour is that a call is made to the xen-create-image command, using the SSH2::exec method (line 1-3). 
This method call should return ONLY (correct me if I'm wrong) after the xen-create-image command is completed, which means that a new virtual machine is ready to be used, and as a consequence, an associated VM configuration file is made available (trialMac.cfg). 
If everything worked fine, there would be no need for the while loop beginning at line 4.

- The **actual** behaviour is that the call to SSH2::exec method, (which launches the xen-create-image command) returns much earlier.. before the virtual machine is created.

In other words, the SSH2::exec method is not able to detect the real completion of the xen-create- image command. 
To overcome this unexpected behaviour, I'm using that while loop beginning at line 4.
 
Is there any chance your code could work in the expected behaviour I described above? 
As usual, if you need any further detail, please let me know.

Thanks again for your support

**Comment by terrafrost** [comment]
Try doing `$sftp->setTimeout(0)` before `$sftp->exec()`. By default the timeout is 10s.

**Comment by narem81** [comment]
I can't believe that :D 
It worked of course! Just FYI I don't seem to have found references about the 10 seconds timeout default in the library documentation, but maybe I missed that. 

Anyway, thank you very much for your help, it has been very appreciated, now everything works as expected!
Keep up with the good work.

**Comment by terrafrost** [comment]
Fix has been merged!
